src/snowflake/connector/connection_diagnostic.py

```diff
@@ -0,0 +1,365 @@
+
+    def __get_certificate(self, host, port=443, timeout=10, host_type="SNOWFLAKE_URL"):
+        context = ssl.create_default_context()
+        context.check_hostname = False
+        context.verify_mode = ssl.CERT_NONE
+        try:
+            conn = socket.create_connection((host, port))
+            sock = context.wrap_socket(conn, server_hostname=host)
+            sock.settimeout(timeout)
+            der_cert = sock.getpeercert(True)
+        except Exception as e:
+            self.test_results[host_type].append(f"{host_type}: {host}:{port}: URL Check: Failed: {e}")
+        finally:
+            sock.close()
+        return ssl.DER_cert_to_PEM_cert(der_cert)
+
+    def __socket_test(self, host, port, host_type="OCSP_RESPONDER"):
+        proxy_url = None
+        proxy_host = self.proxy_host
+        proxy_port = self.proxy_port
+        proxy_user = self.proxy_user
+        proxy_password = self.proxy_password
+        if self.proxy_host is None:
+            if "HTTPS_PROXY" in os.environ.keys():
+                proxy_url = os.environ["HTTPS_PROXY"]
+
+        if proxy_url is not None:
+            parsed = urlparse(proxy_url)
+            proxy_host = parsed.hostname
+            proxy_port = parsed.port
+            proxy_user = parsed.username
+            proxy_password = parsed.password
+
+        try:
+            self.__list_ips(host, host_type=host_type)
+            s = socks.socksocket()
+            s.settimeout(5)
+            if proxy_host is not None:
+                s.set_proxy(socks.HTTP, proxy_host, proxy_port, proxy_user, proxy_password)
+            s.connect((host, int(port)))
+            self.test_results[host_type].append(f"{host_type}: {host}:{port}: URL Check: Connected Successfully")
+        except Exception as e:
+            self.test_results[host_type].append(f"{host_type}: {host}:{port}: URL Check: Failed: {e}")
+            pass
+        finally:
+            s.close()
+
+    def run_post_test(self):
+        if self.whitelist_json is None:
+            try:
+                results = self.cursor().execute(self.whitelist_sql).fetchall()[0][0]
+                results = json.loads(str(results))
+            except Exception as e:
+                logger.warning(f"Unable to do whitelist checks: exception: {e}")
+                pass
+        else:
+            results_file = open(self.whitelist_json)
+            results = json.load(results_file)
+
+        self.whitelist_retrieval_success = True
+        for result in results:
+            host_type = result['type']
+            host = result['host']
+            host_port = result['port']
+            if host_type in ["OUT_OF_BAND_TELEMETRY", "OCSP_RESPONDER"]:
+                if host not in self.ocsp_urls:
+                    self.__socket_test(host, host_port, host_type=host_type)
+            elif host_type in ["STAGE"]:
+                self.__https_host_report(host, port=host_port, host_type="STAGE")
+
+    def __is_privatelink(self):
```

Comment:
If `self.host` doesn't change (after we set it) then we should calculate this in the initalizer

Reply:
This check is actually for the STAGE URL from the system whitelist, I change the parameter names to be more clear.  We check self.host back in run_test.  I can move that test to the initializer.  Honestly, I guess there's no reason we can't have all of run_test in the initializer, then remove run_test all together.  Thoughts?


`    
def run_test(self):
        self.test_results['INITIAL'].append(f"Specified snowflake account: {self.account}")
        self.test_results['INITIAL'].append(f"Host based on specified account: {self.initial_host}")
        self.__check_for_proxies()
        self.__https_host_report(self.host)
        self.ocsp_urls = list(set(self.ocsp_urls))
        for url in self.ocsp_urls:
            self.__socket_test(url, 80)
`